[WIP] Raise error if the user has multiple contexts but no current set

oz123 opened the pull request:
The current behavior is to ignore this and try to connect to `localhost:8080`.
[This is quite confusing, as reported in this issue][1].

This commit improves the UX by telling the users they need to explicitly choose a context, if there are multiple contexts:

```
$ kubectl get pods
error: Multiple contexts found. However, no current-context is set in the configuration
```

[1]: https://github.com/kubernetes/kubectl/issues/936

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This is an attept to improve a confusing error message. Please see the issue linked.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubectl/issues/936

**Special notes for your reviewer**:

I am aware that this PR is still missing test cases. However, I'm publishing this to solicit feedback, and I do not expect this to be merged at this state for the reasons explained below.
With the current patch, if you have not set a context, you see a clear error message. However, you can now only use an editor to fix the issue. That is because this patch disables the ability to use `kubeconfig config set-context`.
```
$ ./kubectl config use-context foo
error: Multiple contexts found. However, no current-context is set in the configuration
```
Users who are a bit familiar with the shell can circumvent the problem:
```
$ echo "current-context: foo" >> ~/.kube/config
```

IMHO, this would be a better interface to exclude the `config` sub-command from producing this error. There are two possible solutions I could think of:
1. Pass the sub-command to the config validator. 
2. Check for the number of contexts in each individual command, e.g. modify all entry points in `kubernetes/vendor/k8s.io/kubectl/pkg/cmd`.

I haven't implemented any yet. It's quite obvious that both approaches make the PR significantly larger. The latter approach makes even larger than the former approach.

Maybe there is another way I have not thought of or not aware of?
I would be happy to get some feedback.

In any case, once I know which approach is preferred I will add the required changes and include tests.

**Does this PR introduce a user-facing change?**:

```release-note
kubectl will now raise an error if the user has multiple contexts but no current set.
Previously, it simply tried to connect to localhost:8080.
```

oz123 committed:
Raise error if the user has multiple contexts but no current set

The current behavior is to ignore this and try to connect to
localhost:8080. [This is quite confusing, as reported in this issue][1].

This commit improves the UX by telling the users they need to explicitly
choose a context in the that multiple context exist:

```
$ kubectl get pods
error: Multiple contexts found. However, no current-context is set in the configuration
```

[1]: https://github.com/kubernetes/kubectl/issues/936

dougsland commented:
looks sane to me. Just wondering why the title is WIP? Still more progress to made related to this RFE? @oz123

oz123 commented:
@dougsland well, it's WIP, because I think there is still some work to do.
For the following reasons:
 * I didn't add any tests
 * I was not sure about the implementation, hence I was thinking out loud in my comment.

I am OK if it's merged like this, but I'm not sure it will be accepted.

dougsland commented:
> @dougsland well, it's WIP, because I think there is still some work to do.
> For the following reasons:
> 
> * I didn't add any tests
> * I was not sure about the implementation, hence I was thinking out loud in my comment.

I agree we might useful to add a test here.
 
> I am OK if it's merged like this, but I'm not sure it will be accepted.

I can't merge patches but let me try to dig here and help moving this one forward.

dougsland commented:
> > @dougsland well, it's WIP, because I think there is still some work to do.
> > For the following reasons:
> > 
> > * I didn't add any tests
> > * I was not sure about the implementation, hence I was thinking out loud in my comment.
> 
> I agree we might useful to add a test here.
> 
> > I am OK if it's merged like this, but I'm not sure it will be accepted.
> 
> I can't merge patches but let me try to dig here and help moving this one forward.

Hi @oz123, 

Okay, I think you didn't get attention deserved here because this is the staging tree under kubernetes git repo (See README from https://github.com/kubernetes/kubernetes/tree/master/staging). 

Please submit your PR to client-go tree: https://github.com/kubernetes/client-go
The tests related to clientcmd are available to look via: validation*.go in the same dir you made the changes.

Next step for me is give a try in your patch. Please let me know if you need any additional help and the new PR url (client-go tree).

fedebongio commented:
/remove-sig api-machinery
/sig cli

dougsland commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -244,11 +244,15 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
 			errlist = append(errlist, err)
 		}
 	}
+
```
just nit: no need to add extra space.

oz123 replied:
Thanks. This is fixed now.

dougsland reviewed (commented):
just double checked, this is the correct repo to send the PR, my bad. The client-go is a mirror.

dougsland commented:
@ahmetb might help here.

oz123 commented:
@douglasland thanks 😊 you beat me to respond. I was trying to comment on this.

corneliusweig commented:
It's not an option to break `kubectl config set-context`. Maybe you can inspect the error list that is returned from the loader in the config subcommand and filter out the error you added here?

oz123 committed:
Allow setting a context when the config does not have one

To enable this sub command, it was necessary to propagate the error
further up the calling chain for inspection. All the subcommands of
config, can now look at the error and decide whether the can continue or
not. As the only command that should still work is "use-context" they
all stop with error if a context is not set.
But this was not enough, since the config is also read while doing the
actual write, it was needed to check for the actually command which was
running. Hence, the signature of `ModifyConfig` and `writeCurrentContext`
was updated. If the use-context subcommand is called the last argument
is set to `true`, this signals that it's OK to ignore the error
`NoConextSet` and continue the execution.
This set of changes displays the error:
```
error: Multiple contexts found. However, no current-context is set in the configuration
```
If running, for example, `kubectl get pods`, but it still allows one to
`get-contexts` and `use-context` quickly from the command line to fix the probelm.

Demo:
```
$ ./kubectl get pods
error: Multiple contexts found. However, no current-context is set in the configuration
$ ./kubectl config get-contexts
CURRENT   NAME               CLUSTER            AUTHINFO         NAMESPACE
          coldsweet          microk8s-cluster   admin-microk8s   coldsweet
          devops             devops-calico      admin
          devops-cilium      devops-cilium      admin-cilium
          microk8s-context   microk8s-cluster   admin-microk8s
$ ./kubectl config use-context devops
Switched to context "devops".
./kubectl get pods
NAME                          READY   STATUS    RESTARTS   AGE
nginx-blue-9cf945578-6gsmz    1/1     Running   0          20d
nginx-green-df5fcc8d5-xk7ww   1/1     Running   0          20d
```

corneliusweig commented on staging/src/k8s.io/client-go/tools/clientcmd/config.go:
```diff
@@ -191,8 +199,14 @@ func ModifyConfig(configAccess ConfigAccess, newConfig clientcmdapi.Config, rela
+
+		err != nil {
+			if isSetContext && errors.Is(err, ErrNoContextSet) {
+
```
empty if blocks are unidiomatic in Go. Try to refactor.

corneliusweig commented on staging/src/k8s.io/client-go/tools/clientcmd/config.go:
```diff
@@ -179,8 +182,13 @@ func ModifyConfig(configAccess ConfigAccess, newConfig clientcmdapi.Config, rela
+
 	if err != nil {
-		return err
+		if isSetContext && errors.Is(err, ErrNoContextSet) {
```
You need to be careful with `errors.Is()`. It returns true if at least one of the reported errors is the respective instance. However, please think about what should happen, if multiple errors are returned in the aggregate.

oz123 replied:
My intent here was to keep the old behavior. Other errors should be reported as before. I added `isSetContext`, so one can still use `set-context`. I tried explaining this in the commit message.
Without this, the check I added caused an error on all the commands.
I will need some time to think of an alternative without `isSetContext`.

corneliusweig replied:
Maybe I missed some detail why this is necessary. However, my main point is to discuss this change in an issue before investing more time into the implementation. It's possible that there is a reason for the current behavior that I don't know.

oz123 replied:
I understand that the change should be discussed in an issue first. This is why I created one and initially sent a very minimal patch. 
Can the discussion for this issue happen on the original issue I opened in the kubectl repository or should I open a new issue in this repository?

corneliusweig reviewed (commented):
Your change has led to some fundamental changes. I think it's better to iterate on it via an issue than a PR to avoid wasting time. The go-to person for these questions are @.soltysh or @.seans3.

soltysh commented on staging/src/k8s.io/client-go/tools/clientcmd/config.go:
```diff
@@ -164,7 +167,7 @@ func NewDefaultPathOptions() *PathOptions {
 // that means that this code will only write into a single file.  If you want to relativizePaths, you must provide a fully qualified path in any
 // modified element.
-func ModifyConfig(configAccess ConfigAccess, newConfig clientcmdapi.Config, relativizePaths bool) error {
+func ModifyConfig(configAccess ConfigAccess, newConfig clientcmdapi.Config, relativizePaths bool, isSetContext bool) error {
```
This is changing external API in a library we provide, you can't do that b/c you're breaking all the existing users.

soltysh reviewed (changes requested):
The rest of the comments from Cornelius still hold.
